packages/@aws-cdk/aws-cloudformation-codepipeline/lib/pipeline-action.ts

```diff
@@ -0,0 +1,318 @@
+     *
+     * @default No capabitilities passed, unless `trustTemplate` is true
+     */
+    capabilities?: CloudFormationCapabilities[];
+
+    /**
+     * Whether the deployed templates are trusted.
+     *
+     * If `true`, the default `role` will have full permissions and the default
+     * `capabilities` will be set to `CloudFormationCapabilities.NamedIAM.`
+     *
+     * If `false`, you will have to set the appropriate IAM permissions and capabilities
+     * manually.
+     *
+     * @default false
+     */
+    trustTemplate?: boolean;
+
+    /**
+     * Input artifact to use for template parameters values and stack policy.
+     *
+     * The template configuration file should contain a JSON object that should look like this:
+     * `{ "Parameters": {...}, "Tags": {...}, "StackPolicy": {... }}`. For more information,
+     * see [AWS CloudFormation Artifacts](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/continuous-delivery-codepipeline-cfn-artifacts.html).
+     *
+     * Note that if you include sensitive information, such as passwords, restrict access to this
+     * file.
+     *
+     * @default No template configuration based on input artifacts
+     */
+    templateConfiguration?: codepipeline.ArtifactPath;
+
+    /**
+     * Additional template parameters.
+     *
+     * Template parameters specified here take precedence over template parameters
+     * found in the artifact specified by the `templateConfiguration` property.
+     *
+     * We recommend that you use the template configuration file to specify
+     * most of your parameter values. Use parameter overrides to specify only
+     * dynamic parameter values (values that are unknown until you run the
+     * pipeline).
+     *
+     * All parameter names must be present in the stack template.
+     *
+     * Note: the entire object cannot be more than 1kB.
+     *
+     * @default No overrides
+     */
+    parameterOverrides?: { [name: string]: any };
+}
+
+/**
+ * Base class for all CloudFormation actions that execute or stage deployments.
+ */
+export abstract class CloudFormationDeploymentAction extends CloudFormationAction {
+    public readonly role: iam.Role;
+
+    constructor(parent: codepipeline.Stage, id: string, props: CloudFormationDeploymentActionCommonProps, configuration: any) {
+        super(parent, id, props, {
+            Capabilities: props.trustTemplate ? [CloudFormationCapabilities.NamedIAM] : props.capabilities,
+            RoleArn: new cdk.Token(() => this.role.roleArn),
+            ParameterOverrides: props.parameterOverrides,
+            TemplateConfiguration: props.templateConfiguration,
+            StackName: props.stackName,
+            ...configuration,
+        });
+
+        if (props.role) {
+            this.role = props.role;
```

Comment:
If you specify your own role, and say you trust the template, the policy doesn't get amended to extend any action on all resources. This seems to contradict the documentation of `trustTemplate`.

Reply:
It does not contradict the documentation of `role`, and it was intended this way (similar to default of `capabilities`). This is a shortcut if you can't be bothered to specify anything. If you can, then you should just spell it out all the way.

Reply:
Ah no, the docs say it adds to the "default role", which is only the role that gets created if you don't specify anything.

Reply:
Feels to me like a warning probably should be emitted...